pkg/validation/commands.go

```diff
@@ -0,0 +1,195 @@
+		return command.Exec.Group
+	case command.Apply != nil:
+		return command.Apply.Group
+	case command.VscodeLaunch != nil:
+		return command.VscodeLaunch.Group
+	case command.VscodeTask != nil:
+		return command.VscodeTask.Group
+	case command.Custom != nil:
+		return command.Custom.Group
+
+	default:
+		return nil
+	}
+}
+
+// validateCommandComponent validates the given exec or apply command, the command should map to a valid container component
+func validateCommandComponent(command v1alpha2.Command, components []v1alpha2.Component) (err error) {
+
+	if command.Exec == nil && command.Apply == nil {
+		return &InvalidCommandError{commandId: command.Id, reason: "should be of type exec or apply"}
+	}
+
+	var commandComponent string
+	if command.Exec != nil {
+		commandComponent = command.Exec.Component
+	} else if command.Apply != nil {
+		commandComponent = command.Apply.Component
+	}
+
+	// must map to a container component
+	isComponentValid := false
+	for _, component := range components {
+		if component.Container != nil && commandComponent == component.Name {
+			isComponentValid = true
+		}
+	}
+	if !isComponentValid {
+		return &InvalidCommandError{commandId: command.Id, reason: "command does not map to a container component"}
+	}
+
+	return
+}
+
+// validateCompositeCommand checks that the specified composite command is valid. The command:
+// 1. should not reference itself via s subcommand
+// 2. should not indirectly reference itself via a subcommand which is a composite command
+// 3. should reference a valid devfile command
+// 4. should have a valid exec sub command
+func validateCompositeCommand(command *v1alpha2.Command, parentCommands map[string]string, devfileCommands map[string]v1alpha2.Command, components []v1alpha2.Component) error {
+
+	// Store the command ID in a map of parent commands
+	parentCommands[command.Id] = command.Id
+
+	if command.Composite == nil {
+		return &InvalidCommandError{commandId: command.Id, reason: "should be of type composite"}
+	}
+
+	// Loop over the commands and validate that each command points to a command that's in the devfile
+	for _, cmd := range command.Composite.Commands {
+		if strings.ToLower(cmd) == command.Id {
+			return &InvalidCommandError{commandId: command.Id, reason: "composite command cannot reference itself"}
+		}
+
+		// Don't allow commands to indirectly reference themselves, so check if the command equals any of the parent commands in the command tree
+		_, ok := parentCommands[strings.ToLower(cmd)]
+		if ok {
+			return &InvalidCommandError{commandId: command.Id, reason: "composite command cannot indirectly reference itself"}
+		}
+
+		subCommand, ok := devfileCommands[strings.ToLower(cmd)]
+		if !ok {
+			return &InvalidCommandError{commandId: command.Id, reason: fmt.Sprintf("the command %q mentioned in the composite command does not exist in the devfile", cmd)}
+		}
+
+		err := validateCommand(subCommand, parentCommands, devfileCommands, components)
```

Comment:
@amisevsk I am not sure I follow the comment here, are you saying this will fail?

exec1
exec2
composite
 - exec1
 - exec2

Reply:
From my read, something like
```yaml
commands:
  - name: composite-1
    composite:
      commands: ["composite-a", "composite-b"]
  - name: composite-a
    composite:
      commands: ["plain-command"]
  - name: composite-b
    composite:
      commands: ["plain-command"]
  - name: plain-command
    exec:
      commandLine: "echo 'hello world'"
```
will be an issue.

Reply:
@amisevsk if we're validating `composite-1` in this scenario, it is perfectly fine IMO, even tho its sub-commands ultimately reference the same exec command `plain-command`

I wrote a test case depicting this scenario and it seems fine!

Reply:
It took me a few tries but I managed to catch the issue I'm describing. Basically, since we're using a flat map to store "visited" command IDs, a command ID that has been processed as part of another branch can cause a false-positive. For example, the references below will fail despite not having any back refs:

![self-ref](https://user-images.githubusercontent.com/16168279/105423065-d9e8a000-5c12-11eb-9ecd-5219a6732567.png)

I pushed a commit with a reproducer [here](https://github.com/amisevsk/devworkspace-api/commit/92d1b19119c61ed44c5a572ccffc21ac1d194056#diff-0321478ca2e7301afce4120f6c3e19ebb5d6401b109284a6cf700dcb1d49476eR1-R19). Excuse the other test cases -- I wanted to make sure my reasoning was right

Reply:
Thanks 🙌 

I was able to solve it by deleting the map entry when one composite command is done being validated.